File: day05/main.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <algorithm>
#include <numeric>
#include <set>
#include <sstream>

auto read() {
    std::fstream ifs(std::string(SRC_DIR)+"/input.txt");
    std::string line;
    std::vector<std::string> result;
    while (std::getline(ifs, line)) {
        if (line.empty())
            continue;
        result.push_back(line);
    }
    return result;
}

/*
    [C]             [L]         [T]
    [V] [R] [M]     [T]         [B]
    [F] [G] [H] [Q] [Q]         [H]
    [W] [L] [P] [V] [M] [V]     [F]
    [P] [C] [W] [S] [Z] [B] [S] [P]
[G] [R] [M] [B] [F] [J] [S] [Z] [D]
[J] [L] [P] [F] [C] [H] [F] [J] [C]
[Z] [Q] [F] [L] [G] [W] [H] [F] [M]
 1   2   3   4   5   6   7   8   9
     */

const std::vector<std::string> original_stacks = {
        {},
        "ZLG",
        "QLRPWFWC",
        "FPMCLGR",
        "LFBWPHM",
        "GCFSVQ",
        "WHJZMQTL",
        "HFSBV",
        "FJZS",
        "MCDPFHBT"
};

void part1() {
    auto stacks = original_stacks;
    for (const auto &line: read()) {
        // move 1 from 5 to 6
        int count = 0, from = 0, to = 0;
        std::sscanf(line.data(), "move %d from %d to %d", &count, &from, &to);
        while (count-- > 0) {
            stacks[to].push_back(stacks[from].back());
            stacks[from].pop_back();
        }
    }

    std::string result;
    std::transform(stacks.begin() + 1, stacks.end(), std::back_inserter(result),
                   [](const std::string &stack) { return stack.back(); });
    std::cout << "Part1: " << result << std::endl;
}


void part2() {
    auto stacks = original_stacks;
    for (const auto &line: read()) {
        // move 1 from 5 to 6
        int count = 0, from = 0, to = 0;
        std::sscanf(line.data(), "move %d from %d to %d", &count, &from, &to);
        auto begin = stacks[from].begin() + stacks[from].size() - count;
        stacks[to].insert(stacks[to].end(), begin, stacks[from].end());
        stacks[from].erase(begin, stacks[from].end());
    }

    std::string result;
    std::transform(stacks.begin() + 1, stacks.end(), std::back_inserter(result),
                   [](const std::string &stack) { return stack.back(); });
    std::cout << "Part2: " << result << std::endl;
}


int main() {
    part1();
    part2();
    return 0;
}

File: day05/src/main.rs
use std::fs::File;
use std::io::{self, BufRead};
use std::path::Path;
use scanf::sscanf;

fn read_lines<P>(filename: P) -> io::Result<io::Lines<io::BufReader<File>>> where P: AsRef<Path> {
    let file = File::open(filename)?;
    Ok(io::BufReader::new(file).lines())
}

fn read() -> Vec<String> {
    let file_path = "input.txt";
    let Ok(lines) = read_lines(file_path) else {
        return vec!();
    };

    return lines.into_iter().map(|x| x.unwrap()).collect();
}

fn parse(line: &str) -> (usize, usize, usize) {
    let mut count :usize = 0;
    let mut from :usize = 0;
    let mut to :usize = 0;

    // move 1 from 5 to 6
    let Ok(_) = sscanf!(&line, "move {} from {} to {}", count, from, to) else {
        panic!("Line does not match");
    };
    (count, from, to)
}

fn part1(input: &Vec<&str>) {

    let mut stacks: Vec<String> = input.iter().map(|&x| x.into()).collect();

    for x in read() {
        let (count,from,to) = parse(&*x);

        for _ in 0..count {
            let c = stacks[from].pop().unwrap();
            stacks[to].push(c);
        }
    }

    let mut result = String::new();
    for i in 1..stacks.len() {
        result.push(stacks[i].chars().last().unwrap());
    }

    println!("Part1: {}", result)
}

fn part2(input: &Vec<&str>) {

    let mut stacks: Vec<String> = input.iter().map(|&x| x.into()).collect();

    for x in read() {
        let (count,from,to) = parse(&*x);

        let len = stacks[from].len();
        let s = stacks[from].drain(len-count..len).collect::<String>();
        stacks[to].push_str(&*s);
    }

    let result = stacks[1..stacks.len()].iter().map(|x| x.chars().last().unwrap()).collect::<String>();
    println!("Part2: {}", result)
}

fn main() {

    /*
   [C]             [L]         [T]
   [V] [R] [M]     [T]         [B]
   [F] [G] [H] [Q] [Q]         [H]
   [W] [L] [P] [V] [M] [V]     [F]
   [P] [C] [W] [S] [Z] [B] [S] [P]
[G] [R] [M] [B] [F] [J] [S] [Z] [D]
[J] [L] [P] [F] [C] [H] [F] [J] [C]
[Z] [Q] [F] [L] [G] [W] [H] [F] [M]
1   2   3   4   5   6   7   8   9
    */

    let stacks = vec![
        "",
        "ZLG",
        "QLRPWFWC",
        "FPMCLGR",
        "LFBWPHM",
        "GCFSVQ",
        "WHJZMQTL",
        "HFSBV",
        "FJZS",
        "MCDPFHBT"
    ];

    part1(&stacks);
    part2(&stacks);
}


